File: example/go-redis/redis/strings.go
package redis

import (
	"context"
	"errors"
	"fmt"
	"time"

	redisv9 "github.com/redis/go-redis/v9"
)

func (r *repository) GET(ctx context.Context, key string) (string, error) {
	result, err := r.client.Get(ctx, key).Result()
	if err != nil {
		if errors.Is(err, redisv9.Nil) {
			return "", fmt.Errorf("key not found")
		}

		return "", err
	}

	return result, nil
}

func (r *repository) SET(ctx context.Context, key, value string) error {
	return r.client.Set(
		ctx,
		key,
		value,
		0).
		Err()
}

func (r *repository) SETEX(ctx context.Context, key, value string, expiration time.Duration) error {
	return r.client.Set(
		ctx,
		key,
		value,
		expiration,
	).Err()
}

func (r *repository) TTL(ctx context.Context, key string) (time.Duration, error) {
	ttl, err := r.client.TTL(ctx, key).Result()
	if err != nil {
		return 0, err
	}
	return ttl, nil
}

File: example/rust-redis/src/redis/strings.rs

use anyhow::{anyhow, Result};
use redis::AsyncCommands;

use super::redis::Repository;

impl Repository {
    pub async fn get(&self, key: &str) -> Result<String> {
        let mut conn = self.client.get_multiplexed_async_connection().await?;

        let result: Option<String> = conn.get(key).await?;

        match result {
            Some(value) => Ok(value),
            None => Err(anyhow!("key not found")),
        }
    }

    pub async fn set(&self,key: &str, value: &str) -> Result<()> {
        let mut conn = self.client.get_multiplexed_async_connection().await?;
        conn.set::<_, _, ()>(key, value).await?;
        Ok(())
    }

    pub async fn setex(&self,key: &str,value: &str,expiration: u64,) -> Result<()> {
        let mut conn = self.client.get_multiplexed_async_connection().await?;
        conn.set_ex::<_, _, ()>(
            key,
            value,
            expiration,
        )
        .await?;

        Ok(())
    }

    pub async fn ttl(&self, key: &str) -> Result<i64> {
        let mut conn = self.client.get_multiplexed_async_connection().await?;
        let ttl: i64 = conn.ttl(key).await?;
        Ok(ttl)
    }
}

